### PR TITLE
osd: Let the osd kowns who kills me

### PR DESCRIPTION
**Pull request by songbaisen**
OSD: Let the osd kown who kills me

Fixes:#13576
Signed-off-by:song baisen <song.baisen@zte.com.cn>

### CODE REVIEW

File: src/global/signal_handler.cc [hunk 1]
```diff
@@ -288,12 +316,12 @@ void SignalHandler::register_handler(int signum, signal_handler_t handler, bool
   assert(signum >= 0 && signum < 32);
 
   safe_handler *h = new safe_handler;
-
+  h->is_siginfo = false;
```
Comment (ifed01) [inline src/global/signal_handler.cc]: Probably it's better to use ctor for that.

File: src/global/signal_handler.cc [hunk 2]
```diff
@@ -172,6 +172,9 @@ struct SignalHandler : public Thread {
   struct safe_handler {
     int pipefd[2];  // write to [1], read from [0]
     signal_handler_t handler;
+	signal_siginfo_handler_t siginfo_handler;
```
Comment (ifed01) [inline src/global/signal_handler.cc]: Broken indention here and below

File: src/global/signal_handler.cc [hunk 3]
```diff
@@ -246,10 +249,20 @@ struct SignalHandler : public Thread {
 	lock.Lock();
 	for (unsigned signum=0; signum<32; signum++) {
 	  if (handlers[signum]) {
-	    r = read(handlers[signum]->pipefd[0], &v, 1);
-	    if (r == 1) {
-	      handlers[signum]->handler(signum);
-	    }
+	   if(true==handlers[signum]->is_siginfo)
+        {
+            r = read(handlers[signum]->pipefd[0], &v, 1);
```
Comment (ifed01) [inline src/global/signal_handler.cc]: You may want to put this line out of this 'if' block as it's the same for both true/false cases

File: src/global/signal_handler.cc [hunk 4]
```diff
@@ -269,9 +282,19 @@ struct SignalHandler : public Thread {
     int r = write(handlers[signum]->pipefd[1], " ", 1);
     assert(r == 1);
   }
-
+  
+  void quenue_siginfo_signal(int signum,siginfo_t * siginfo,void * content)
```
Comment (ifed01) [inline src/global/signal_handler.cc]: quenue - > queue typo?

File: src/osd/OSD.cc [hunk 5]
```diff
@@ -1652,6 +1652,27 @@ void OSD::handle_signal(int signum)
...
 
+void OSD::handle_signal_info(int signum,siginfo_t * siginfo, void * content)
+{
+	assert(signum == SIGINT || signum == SIGTERM || signum == SIGABRT);
+	derr << "*** Got signal " << sys_siglist[signum] << " ***" << dendl;
+	derr << "*** PID: "<<siginfo->si_pid<<"***UID: "<<siginfo->si_uid<<dendl;
+
+	//suicide(128 + signum);
+
+	if (signum == SIGABRT) {
```
Comment (ifed01) [inline src/osd/OSD.cc]: Do we need to register signal handler for SIGABRT?

File: src/osd/OSD.cc [hunk 6]
```diff
@@ -1652,6 +1652,27 @@ void OSD::handle_signal(int signum)
   shutdown();
```
Comment (ifed01) [inline src/osd/OSD.cc]: we don't need this assertion any more, do we? And we probably lack SIGHUP here

File: src/osd/OSD.cc [hunk 7]
```diff
@@ -1652,6 +1652,27 @@ void OSD::handle_signal(int signum)
...
+{
+	assert(signum == SIGINT || signum == SIGTERM || signum == SIGABRT);
+	derr << "*** Got signal " << sys_siglist[signum] << " ***" << dendl;
+	derr << "*** PID: "<<siginfo->si_pid<<"***UID: "<<siginfo->si_uid<<dendl;
+
+	//suicide(128 + signum);
+
+	if (signum == SIGABRT) {
+	    //if we die unexpectedly, try to leave some traces.
+		printf(
```
Comment (ifed01) [inline src/osd/OSD.cc]: Why not using output to derr here?

**Comment by songbaisen** [comment]
@ifed01 thank you for your good advice.

**Comment by xiaoxichen** [comment]
what's the use case for this PR? Ceph will not kill itself , so the killer should be outside ceph, might be administrator or some 3rd party monitor tool?

And pls stash some commit that make every commit meaningful.

**Comment by songbaisen** [comment]
@xiaoxichen yes,ceph will not kill himself. But when use ceph with other sofrware , The othe software may kill or terminate ceph.But we should konw who kill ceph.So that we can find the reason.

File: src/ceph_osd.cc [hunk 8]
```diff
@@ -68,6 +68,12 @@ void handle_osd_signal(int signum)
     osd->handle_signal(signum);
 }
 
+void handle_osd_signal_info(int signum,siginfo_t * siginfo, void * content)
+{
+  if (osd)
+    osd->handle_signal_info(signum,siginfo,content);
```
Comment (runsisi) [inline src/ceph_osd.cc]: need a space after the comma

**Comment by xiaoxichen** [comment]
This requirement looks a bit strange for me, waiting for other's input.....I think the administrator should be very clear about who(which program) has the privileged to kill other process.

**Comment by songbaisen** [comment]
@xiaoxichen  If the dministrator  kowns all the software in the compute that is ok.But in our environment.When install ceph with the openstack.We find the osd usually killed by the openstack guard thread.This problem we use a lot time to know. And we develop a ceph web like calamari,we  use salt-minion,we also find a bug for that software,that kill ceph osd or minitor.And we also waster a lot time to kown the reason.So i  decide to let the osd and mon kowns who kill it.

**Comment by songbaisen** [comment]
@loic-bot  Hi.You're working so hard for ceph.HAHA.

**Comment by hjwsm1989** [comment]
+1, i think it's useful if we know who killed osd or mon, it will save a lot time for administrator.

**Comment by liewegas** [comment]
I like the goal here.  But I think we can make this *much* simpler by just modifying global/signal_handler.cc to print out the process that sent the signal at debug level 0.  It means we'll get a "received SIGHUP from logrotate" or similar message regularly in the log, but I think that's actually a good thing.  And it means that all the handlers themselves don't have to change--just global/signal_handler.cc.

Does that sound reasonable?

**Comment by songbaisen** [comment]
@liewegas  Thank you sir! That is a good idea!

**Comment by athanatos** [comment]
Closing for now.  Feel free to open a new one with the new approach (or just reopen this one if you think I shouldn't have closed it)!

**Comment by songbaisen** [comment]
@liewegas  @athanatos :OK,I will add this in another new approach .

**Pull request closed without merging**